src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -483,26 +495,67 @@ private void Heapify()
             int lastNodeIndex = GetLastNodeIndex();
             int lastParentWithChildren = GetParentIndex(lastNodeIndex);
 
-            for (int index = lastParentWithChildren; index >= 0; --index)
+            if (_comparer == null)
             {
-                MoveDown(_nodes[index], index);
+                for (int index = lastParentWithChildren; index >= 0; --index)
+                {
+                    MoveDownDefaultComparer(_nodes[index], index);
+                }
+            }
+            else
+            {
+                for (int index = lastParentWithChildren; index >= 0; --index)
+                {
+                    MoveDownCustomComparer(_nodes[index], index);
+                }
             }
         }
 
         /// <summary>
         /// Moves a node up in the tree to restore heap order.
         /// </summary>
-        private void MoveUp((TElement Element, TPriority Priority) node, int nodeIndex)
+        private void MoveUpDefaultComparer((TElement Element, TPriority Priority) node, int nodeIndex)
         {
             // Instead of swapping items all the way to the root, we will perform
             // a similar optimization as in the insertion sort.
 
+            Debug.Assert(_comparer is null);
+
             while (nodeIndex > 0)
             {
                 int parentIndex = GetParentIndex(nodeIndex);
                 (TElement Element, TPriority Priority) parent = _nodes[parentIndex];
 
-                if (Comparer.Compare(node.Priority, parent.Priority) < 0)
+                if (Comparer<TPriority>.Default.Compare(node.Priority, parent.Priority) < 0)
```

Comment:
To Consider: should we special case reference types?

Reply:
Yes, but I think that can be done without changes here, and rather just in the ctor by ensuring reference types always have a non-null comparer.